### ISSUE TITLE
Use NEURON's Random123 for connectivity and cell locations

### ISSUE THREAD

**Issue by salvadord** [Neurosim-lab/netpyne#89]

**Comment by salvadord** [comment]
already using noiseFromRandom123() for NetStims

**Comment by salvadord** [comment]
ramcdougal [2:15 PM] 
so... you probably shouldn't use Python's `random` module for parallel network models; it doesn't have a good way of generating independent random streams where the randomness is independent of the number of processors you're using

sidmitra [2:26 PM] 
That is indeed a concern. I found this on stackexchange - https://scicomp.stackexchange.com/questions/24916/how-do-i-reliably-generate-random-numbers-in-python-distributed-across-multiple

[2:26] 
It suggests that Randomstate be used instead.

[2:26] 
https://pypi.python.org/pypi/randomstate/1.10.1

billl [2:29 PM] 
NEURON(namely hines) has put lot of effort into this issue so well supported

salvadord [3:06 PM] 
I did have issues but think solved them — tested results reproducible and independent of num cores — but will check again, thx

**Comment by salvadord** [comment]
Currently randomizer seed is the sum of the conn seed plus  the postsyn gid (convergence), presyn gid (divergence), or 1st gids of pre and post pops (probability). However this leads to several issues:
- sums of gids may coincide leading to same random stream (eg. pregid=1, postgid=2 and pregid=2,postgid=1)
- convergence conn rules to same postsyn gid will generate same random streams. This shouldn't be an issue if the presyn pop is different ?
- divergence conn rules from same presyn gid will generate same random streams.
- prob conn rules will generate same random stream for same pre and post gids

**Comment by salvadord** [comment]
salvadord [1 hour ago] 
ramcdougal: so just checked that random conns are independent of num of cores; basically different postsyn cells get assigned to different nodes, and randomizer seed is function of postsyn cell

salvadord [1 hour ago] 
but you are saying also maybe the random streams are not completely independent when using py’s random?

salvadord [45 minutes ago] 
but don’t see how this problem is dependent on num of nodes, instead of simply an issue with py’s random ?

ramcdougal [19 minutes ago] 
short answer is that it probably doesn't matter

ramcdougal [15 minutes ago] 
mathematical answer is that: if you're using the Mersenne Twister (Python's default prng) and you set, say, seed(1) vs seed(2) what you're doing is jumping into different places in a single random stream. (There's a scrambling that happens that makes the locations not adjacent, but it's still the same stream... pull enough bits and they'll align.)

salvadord [14 minutes ago] 
aah ic, thx makes sense!

ramcdougal [13 minutes ago] 
This is why NEURON uses Random123 which is a cryptographic quality generator (Mersenne Twister is not) that creates independent random streams based on 3 ints. The usual way to do this is to have a global seed and then a per cell stream id.

ramcdougal [11 minutes ago] 
For the comp bio course here, I recommended the `randomstate` module which also provides mathematically independent random streams

ramcdougal [11 minutes ago] 
(not part of standard Python; has to be installed separately)

salvadord [9 minutes ago] 
thanks cool I’ll check randomstate, maybe can include package within netpyne

ramcdougal [9 minutes ago] 
(and so yes: doing it the way you're doing it is free of any issues with how many compute nodes... but doesn't avoid the issue with the random module itself)

salvadord [8 minutes ago] 
also I just realized that if seed depends on global seed + postsyn gid, if you create 2 sets of conns to same postsyn cell, they will generate same random stream, right?

ramcdougal [8 minutes ago] 
sounds like it

ramcdougal [7 minutes ago] 
there may not be a reason to use `randomstate` since NEURON has `Random123` to solve this exact problem

ramcdougal [6 minutes ago] 
and so then: that takes multiple identifiers to select the stream: so you have a seed, a pre-cell, and a post-cell

salvadord [6 minutes ago] 
random123 - right, I use for netstims etc, but avoided for conn since prefer not using Vectors()

ramcdougal [5 minutes ago] 
also if I'm understanding your current approach, then the stream being fed to global seed 1's cell 2 is the same as that going to global seed 2's cell 1... so your separate simulations wouldn't be independent

salvadord [5 minutes ago] 
multiple, pre, post - ic, although for example to select the random presyn cells for convergence connectivity could only use the post gid

ramcdougal [4 minutes ago] 
Vectors? **checks the python documentation that exists now**

ramcdougal [4 minutes ago] 
nope. confused. why Vectors? https://www.neuron.yale.edu/neuron/static/py_doc/programming/math/random.html#Random.Random123

salvadord [3 minutes ago] 
understand approach seed 1's cell2 - yes! I also realized that and have an issue in github to fix

salvadord [1 minute ago] 
vectors - I had only used Random123 with h.Vectors(), but will check py doc

salvadord [1 minute ago] 
thanks!

**Comment by salvadord** [comment]
ramcdougal [24 minutes ago] 
oh: yeah, you can fill a Vector from a NEURON random stream, but... you can also pull a single number at a time, much like you can from Python's Random

salvadord [22 minutes ago] 
great, guess doesn’t have equivalent of random.sample() ?

ramcdougal [17 minutes ago] 
no... it's either fill a vector or something like `[r.repick() for i in range(sample_size)]`

ramcdougal [15 minutes ago] 
arguably filling a Vector is basically the equivalent of random.sample, but admittedly slightly annoying to work with in Python. Is building the random list NetPyNE's responsibility or the user's? If the user's, you could provide a `netpyne.sample(r, n)` or something

**Comment by salvadord** [comment]
rationale for `rand.Random123(sim.id32('conn_str'), sim.id32('%d%d%d%d'%(len(preCellsTags), len(postCellsTags), sum(preCellsTags), sum(postCellsTags)), sim.cfg.seeds['conn']))`:

- sim.id32('conn_str'): different random streams for 'probability' vs 'convergence' vs 'divergence (in cases when have same set of pre and postsyn cells

- sim.id32('%d%d%d%d'%(len(preCellsTags), len(postCellsTags), sum(preCellsTags), sum(postCellsTags)): different random streams for different subsets of pre and post cells; also different if reverse pre and post; can have duplicate streams if sum of gids coincide (eg. 1+7 = 2+6)

-  sim.cfg.seeds['conn']: different but reproducible random streams for each conn seed

**Comment by salvadord** [comment]
salvadord [4:10 PM]
are you aware of any advantage of NEURON’s h.Random() over python’s RandomState?

ramcdougal [4:12 PM]
yes
RandomState is Mersenne Twister

salvadord [4:18 PM]
ok so h.Random() arguably better for large-scale sims right?
guess this paper talks about that but haven’t read: http://www.thesalmons.org/john/random123/papers/random123sc11.pdf
its just that performance analysis showed that seeding h.Random() for each pair of pre- and post-gids was really time-consuming, so was looking at alternatives

ramcdougal [9:31 AM]
hopefully you're using the h.Random.Random123?

salvadord [9:31 AM]
y

ramcdougal [9:31 AM]
so basically, yeah Salmon et al addresses the algorithm we use
Mersenne Twister isn't cryptographic qualkity, 600-something observations gives you a predictable sequence

salvadord [9:37 AM]
ok thanks, I’ll stick with Random123 then

**Comment by salvadord** [comment]
Optimized performance by:
- using a single seeding for each conn rule
- seeding based on 1) hash of all preGids, 2) hash of all postGids, 3) global conn seed
- hashing based on array-based fast method
- use vec.setrand(r) to create vector of values; then convert to list
- random values then mapped to preGid,postGid in dict

`   
def hashList(obj):
    return int(hashlib.md5(array.array('L', obj)).hexdigest()[0:8],16)

def generateRandsPrePost(self, pre, post):
    from .. import sim

    # initialize randomizer using unique hash of pre and post gids and global conn seed
    self.rand.Random123(sim.hashList(sorted(pre)), sim.hashList(sorted(post)), sim.cfg.seeds['conn'])  #
    
    # obtain rand value for pre,post pairs
    lenPre = len(pre)
    lenPost = len(post)
    vec = sim.h.Vector(lenPre*lenPost)  # create Vector
    self.rand.uniform(0,1)  # set unfiform distribution
    vecList = list(vec.setrand(self.rand))  # fill in vector 
    allRands = {(preGid,postGid): vecList[(ipre*lenPost)+ipost] for ipre,preGid in enumerate(pre) for ipost,postGid in enumerate(post)}  # convert to dict

    return allRands
`
